### PR TITLE
feat(unlock-app) - Checkout url

### PR DESCRIPTION
**Pull request by kalidiagne**
# Description
In order to give the user the ability to easily build the checkout URL we add the page to simplify it:

## Overview video

https://user-images.githubusercontent.com/20865711/197238216-ff17bdbb-c248-45a2-a9ce-066d4ac6c47f.mov

**Remove from list**

remove-lock-form-list

**Copy link**

copy-link-builder

## Screens 

**Metadata** 

Screenshot 2022-10-21 at 17 27 23

Screenshot 2022-10-21 at 17 26 58

**Preview with valid lock** 
Screenshot 2022-10-20 at 19 55 26

**Preview without lock** 
Screenshot 2022-10-20 at 19 56 43

**Lock selection**
Screenshot 2022-10-20 at 19 57 29

Screenshot 2022-10-20 at 19 57 43

**Lock list & preview**
Screenshot 2022-10-20 at 19 59 14

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by kalidiagne** [comment]
Form generated from zod-schema

Screenshot 2022-10-13 at 15 10 40

Screenshot 2022-10-13 at 15 10 26

File: unlock-app/package.json [hunk 1]
```diff
@@ -85,7 +85,8 @@
     "validator": "13.7.0",
     "walletlink": "2.5.0",
     "xstate": "4.32.1",
-    "zod": "3.17.10"
+    "zod": "3.17.10",
+    "zod-to-json-schema": "^3.18.1"
```
Comment (julien51) [inline unlock-app/package.json:89]: Remove the ^

File: unlock-app/src/components/interface/locks/CheckoutUrl/description.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import { PaywallConfig, PaywallConfigLock } from '~/unlockTypes'
+
+type Description = Partial<
+  Record<keyof PaywallConfig | keyof PaywallConfigLock, string>
+>
+
+export const PaywallDescriptions: Description = {
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/description.ts:7]: Do we have any way to add this to the zod schema itself?

Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/description.ts:7]: Is there a way to include these in the zod schema?

Comment (kalidiagne) [reply unlock-app/src/components/interface/locks/CheckoutUrl/description.ts:7]: yes! ill move it now into zod!

Comment (kalidiagne) [reply unlock-app/src/components/interface/locks/CheckoutUrl/description.ts]: https://github.com/unlock-protocol/unlock/pull/9993

**Comment by kalidiagne** [comment]
Temp removed `ToggleSwitch` with a basic `input`.
We need to update the `ToggleSwitch` to pass ref. 

@julien51 if all ok you can merge and ill update it later

**Comment by julien51** [comment]
Let me give a look and change a few things before merging!

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import useClipboard from 'react-use-clipboard'
+import { ToastHelper } from '~/components/helpers/toast.helper'
+import FileSaver from 'file-saver'
+
+interface CheckoutPreviewProps {
+  paywallConfig?: PaywallConfig
+}
+
+const onDownloadJson = (paywallConfig: PaywallConfig) => {
+  const fileName = 'checkout-config.json'
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx:17]:
```suggestion
  const fileName = 'paywall-config.json'
```

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
+import { Button } from '@unlock-protocol/ui'
+import { useEffect, useState } from 'react'
+import { Checkout } from '~/components/interface/checkout/alpha/Checkout'
+import { selectProvider } from '~/hooks/useAuthenticate'
+import { useCheckoutCommunication } from '~/hooks/useCheckoutCommunication'
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx:5]: I don't think we need this one? It's for communication in an iframe...

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    type: 'application/json',
+  })
+
+  FileSaver.saveAs(fileToSave, fileName)
+}
+export const CheckoutPreview = ({ paywallConfig }: CheckoutPreviewProps) => {
+  const [checkoutUrl, setCheckoutUrl] = useState('')
+  const config = useConfig()
+  // Fetch config from parent in iframe context
+  const communication = useCheckoutCommunication()
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx:30]:
```suggestion
```

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+  const injectedProvider =
+    communication.providerAdapter || selectProvider(config)
+
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx:34]:
```suggestion
  const injectedProvider = selectProvider(config)

```

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  const checkoutRedirectURI = ''
+
+  const hasLocks = Object.entries(paywallConfig?.locks ?? {})?.length > 0
+
+  useEffect(() => {
+    const url = new URL(`${window.location.origin}/alpha/checkout`)
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx:40]:
```suggestion
    const url = new URL(`${window.location.origin}/checkout`)
```

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  const buttonVariant = hasLocks ? 'primary' : 'outlined-primary'
+
+  return (
+    <div className="flex items-center justify-center w-full py-10 bg-gray-300 rounded-3xl">
+      <div className="flex items-center justify-center w-full max-w-lg">
+        {paywallConfig && (
+          <div className="flex flex-col items-center w-full gap-4">
+            <Checkout
+              injectedProvider={injectedProvider as any}
+              communication={communication}
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx:65]: we also don't need this here I think

**Review by julien51 — COMMENTED** [review]
I realize we are missing the configuration for the redirect at the end...  since it is not technically part of the paywall config let's add it as a another tab?

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+              paywallConfig={paywallConfig as any}
+              redirectURI={
+                checkoutRedirectURI ? new URL(checkoutRedirectURI) : undefined
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx:68]: HA! I realize we never spoke about the redirect URL configuration

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/DynamicForm.tsx [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+  integer: TextInput,
+  boolean: BooleanInput,
+}
+
+const TypeMap: Record<string, string> = {
+  string: 'text',
+  integer: 'number',
+}
+
+const camelCaseToText = (text: string) => {
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/DynamicForm.tsx:97]: Hum, unfortunately this is not the right approach here. We need to have a mapping of `label` like we have a mapping of `description`.

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/DynamicForm.tsx [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+              )
+            }
+
+            if (!Component) return null
+
+            return (
+              <div className="flex flex-col gap-2" key={index}>
+                <Component
+                  props={props}
+                  label={camelCaseToText(label)}
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/DynamicForm.tsx:196]: See my comment above...

**Review by julien51 — COMMENTED** [review]
Let's please add a way to "edit" any lock that was previously added (vs. deleting and adding it again)
Also, please let's change the label so we have a mapping, this will be looking better and allow us to make better names as well as i18n after that!

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx (hunk 8 above)
Comment (kalidiagne) [reply unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx:65]: is required from `Checkout` component as prop

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx (hunk 9 above)
Comment (kalidiagne) [reply unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx]: we get this value actually, maybe we can move to the top position instead:

Screenshot 2022-10-24 at 12 13 30

Comment (julien51) [reply unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx]: > we get this value actually, maybe we can move to the top position instead:

Yes let's do this please.

**Comment by kalidiagne** [comment]
> Let's please add a way to "edit" any lock that was previously added (vs. deleting and adding it again) Also, please let's change the label so we have a mapping, this will be looking better and allow us to make better names as well as i18n after that!

https://user-images.githubusercontent.com/20865711/197532196-5b261135-5b2f-4670-a2d5-6a9e7da18b91.mov

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx (hunk 8 above)
Comment (julien51) [reply unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx:66]: Yeah I think we should make it optional...

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/DynamicForm.tsx [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+  messageToSign: 'Message to sign',
+  pessimistic: 'Pessimistic',
+  hideSoldOut: 'Hide Sold-out',
+  type: 'Type',
+  placeholder: 'Placeholder text',
+  defaultValue: 'Default value',
+  required: 'Required',
+  public: 'Shown as public',
+  recurringPayments: 'Recurring frequency',
+  minRecipients: 'Min. Quantity',
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/DynamicForm.tsx:23]:
```suggestion
  minRecipients: 'Minimum number of recipients',
```

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/DynamicForm.tsx [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+  maxRecipients: 'Max. Quantity',
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/DynamicForm.tsx:24]:
```suggestion
  maxRecipients: 'Maximum number of Recipients',
```

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/DynamicForm.tsx [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+  password: 'Password verification',
```
Comment (julien51) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/DynamicForm.tsx:25]:
```suggestion
  password: 'Requires a password',
```

**Review by julien51 — APPROVED** [review]
Please remove the iframe communication stuff and then LG!

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx (hunk 9 above)
Comment (kalidiagne) [reply unlock-app/src/components/interface/locks/CheckoutUrl/elements/CheckoutPreview.tsx]: moved to first position for the section

Screenshot 2022-10-24 at 16 03 25

File: unlock-app/src/components/interface/locks/CheckoutUrl/elements/LocksForm.tsx [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+            first: 1000,
+            where: {
+              lockManagers_contains: [account!],
+            },
+          },
+          {
+            networks: [`${network!}`],
+          }
+        )) ?? []
+      )
```
Comment (searchableguy) [inline unlock-app/src/components/interface/locks/CheckoutUrl/elements/LocksForm.tsx:108]: nit: move it into a variable and then setLocksByNetwork.

**Review by searchableguy — APPROVED** [review]
LGTM. 
I think we could refactor some parts to use `useList` from `react-use` which provides a way to manage list state.

**Pull request merged**